src/server/bucket_guard.h

```diff
@@ -0,0 +1,28 @@
+// Copyright 2024, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#pragma once
+#include "util/fibers/synchronization.h"
+
+namespace dfly {
+// Helper class used to guarantee atomicity between serialization of buckets
+template <typename T> class BucketSerializationGuard {
+ public:
+  explicit BucketSerializationGuard(T* enclosing) : enclosing_(enclosing) {
```

Comment:
instead of template T I would create a stuct that holds 
bucket_ser_cond and bucket_ser_in_progress_, add this stuct as member of snapshot and streamer and use BucketSerializationGuard which will get this struct on constructor.